File: homework/src/main/java/com/jlm/homework/entity/ExerciseBookEntity.java
package com.jlm.homework.entity;

import com.alibaba.cloud.commons.lang.StringUtils;
import com.alibaba.nacos.shaded.com.google.gson.Gson;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.persistence.*;
import lombok.Data;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 练习册实体类
 * 对应数据库表 exercise_book
 */
@Data
@Entity
@Table(name = "exercise_book")
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class ExerciseBookEntity {
    /**
     * 主键ID
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    /**
     * 练习册标题
     */
    @Column(name = "title", nullable = false, length = 200)
    private String title;

    /**
     * 练习册描述
     */
    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    /**
     * 学科名称（保留用于显示）
     */
    @Column(name = "subject", length = 50)
    private String subject;

    /**
     * 学科ID
     */
    @Column(name = "subject_id")
    private Long subjectId;

    /**
     * 年级名称（保留用于显示）
     */
    @Column(name = "grade", length = 20)
    private String grade;

    /**
     * 年级ID
     */
    @Column(name = "grade_id")
    private Long gradeId;

    /**
     * 班级ID（单个，保留向后兼容）
     */
    @Column(name = "class_id")
    private Long classId;

    /**
     * 班级ID列表（JSON格式存储）
     */
    @Column(name = "class_ids", columnDefinition = "JSON")
    private String classIds;

    /**
     * 班级名称列表（JSON格式存储）
     */
    @Column(name = "class_names", columnDefinition = "JSON")
    private String classNames;

    /**
     * 难度等级 (1-5)
     */
    @Column(name = "difficulty_level")
    private Integer difficultyLevel = 1;

    /**
     * 创建者ID
     */
    @Column(name = "creator_id")
    private Long creatorId;

    /**
     * 学校ID（租户标识）
     */
    @Column(name = "school_id")
    private Long schoolId;

    /**
     * 状态 (ACTIVE, INACTIVE, DELETED)
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20)
    private ExerciseBookStatus status = ExerciseBookStatus.ACTIVE;

    /**
     * 练习册图片列表（JSON格式存储）
     */
    @Column(name = "images", columnDefinition = "JSON")
    private String images;

    /**
     * 创建时间
     */
    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime createdAt;

    /**
     * 更新时间
     */
    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime updatedAt;
    @Transient
    private List<Long> classIdList;
    @Transient
    private List<String> classNameList;
    @Transient
    private List<String> imageUrls;

    public List<Long> getClassIdList() {
        if(StringUtils.isNotEmpty(classIds)){
            Gson gson = new Gson();
            Long[] ids= gson.fromJson(classIds,Long[].class);
            return Arrays.asList(ids);
        }else {
            return new ArrayList<>();
        }

    }
    public List<String> getClassNameList() {
        if(StringUtils.isNotEmpty(classNames)){
            Gson gson = new Gson();
            String[] ids= gson.fromJson(classNames,String[].class);
            return Arrays.asList(ids);
        }else {
            return new ArrayList<>();
        }

    }    // 单例ObjectMapper，避免重复创建
    private static final ObjectMapper objectMapper = new ObjectMapper();

    // 构造函数
    public ExerciseBookEntity() {
    }

    public ExerciseBookEntity(String title, String description, String subject,Long subjectId,String grade,Long gradeId,Long classId, Integer difficultyLevel, Long creatorId, Long schoolId,ExerciseBookStatus status) {
        this.title=title;
        this.description=description;
        this.subject=subject;
        this.subjectId=subjectId;
        this.grade=grade;
        this.gradeId=gradeId;
        this.classId= classId; // 保留向后兼容
        this.difficultyLevel=difficultyLevel==null?1:difficultyLevel;
        this.creatorId=creatorId;
        this.schoolId=schoolId;
        this.status=status == null ?  ExerciseBookStatus.ACTIVE:status;

    }


    /**
     * 获取图片列表
     */
    public List<ExerciseBookImage> getImageList() {
        try {
            if (images == null || images.isBlank()) {
                return new ArrayList<>();
            } else {
                return objectMapper.readValue(
                        images,
                        objectMapper.getTypeFactory().constructCollectionType(List.class, ExerciseBookImage.class)
                );
            }
        } catch (IOException e) {
            // 在Java中通常会记录异常日志
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    /**
     * 设置图片列表
     */
    public void setImageList(List<ExerciseBookImage> imageList) {
        try {
            this.images = objectMapper.writeValueAsString(imageList);
        } catch (IOException e) {
            // 在Java中通常会记录异常日志
            e.printStackTrace();
        }
    }
    /**
     * 创建带图片的练习册实体
     */
    public ExerciseBookEntity withImages(List<ExerciseBookImage> imageList) {
        if (imageList.isEmpty()) {
            this.images = null;
        }
        try {
            this.images = objectMapper.writeValueAsString(imageList);
        } catch (IOException e) {}
        return this;
    }
    /**
     * 从URL列表创建图片信息
     */
    public List<ExerciseBookImage> createImagesFromUrls(List<String> urls) {
        List<ExerciseBookImage>  imageList = new ArrayList<>();
        for (int i = 0; i < urls.size(); i++) {
            ExerciseBookImage  image = new ExerciseBookImage();
            image.setOrder(i);
            if(i==0){
                image.setType("cover");
            }else{
                image.setType("content");
            }

            image.setUrl(urls.get(i));
            imageList.add(image);
        }
        return imageList;
    }
    public ExerciseBookEntity withClassIds(List<Long> classIdList) {
        if (classIdList.isEmpty()) {
            this.setClassIds(null);
        } else {
            String classIds=null;
            try {
                classIds =objectMapper.writeValueAsString(classIdList);
            } catch ( Exception e) {
                classIds =null;
            }
            this.setClassIds(classIds);
        }
        return this;
    }
    public ExerciseBookEntity withClassIdsAndNames(List<Long> classIds, List<String> classNames) {
        if (classIds.isEmpty()) {
            this.setClassIds(null);
        } else {
            String classIdsStr=null;
            try {
                classIdsStr =objectMapper.writeValueAsString(classIds);
            } catch ( Exception e) {
                classIdsStr =null;
            }
            this.setClassIds(classIdsStr);
        }

        if (classNames.isEmpty()) {
            this.setClassNames(null);
        } else {
            String classNamesStr=null;
            try {
                classNamesStr =objectMapper.writeValueAsString(classNames);
            } catch ( Exception e) {
                classNamesStr =null;
            }
            this.setClassNames(classNamesStr);
        }
        return this;
    }




}

File: homework/src/main/kotlin/com/jlm/homework/entity/ExerciseBookEntity.kt
package com.jlm.homework.entity

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonProperty
import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.LocalDateTime

/**
 * 练习册实体类
 * 对应数据库表 exercise_book
 */
@Entity
@Table(name = "exercise_book")
data class ExerciseBookEntity(
    /**
     * 主键ID
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    val id: Long? = null,

    /**
     * 练习册标题
     */
    @Column(name = "title", nullable = false, length = 200)
    val title: String? = null,

    /**
     * 练习册描述
     */
    @Column(name = "description", columnDefinition = "TEXT")
    val description: String? = null,

    /**
     * 学科名称（保留用于显示）
     */
    @Column(name = "subject", length = 50)
    val subject: String? = null,

    /**
     * 学科ID
     */
    @Column(name = "subject_id")
    val subjectId: Long? = null,

    /**
     * 年级名称（保留用于显示）
     */
    @Column(name = "grade", length = 20)
    val grade: String? = null,

    /**
     * 年级ID
     */
    @Column(name = "grade_id")
    val gradeId: Long? = null,

    /**
     * 班级ID（单个，保留向后兼容）
     */
    @Column(name = "class_id")
    val classId: Long? = null,

    /**
     * 班级ID列表（JSON格式存储）
     */
    @Column(name = "class_ids", columnDefinition = "JSON")
    val classIds: String? = null,

    /**
     * 班级名称列表（JSON格式存储）
     */
    @Column(name = "class_names", columnDefinition = "JSON")
    val classNames: String? = null,

    /**
     * 难度等级 (1-5)
     */
    @Column(name = "difficulty_level")
    val difficultyLevel: Int? = 1,

    /**
     * 创建者ID
     */
    @Column(name = "creator_id")
    val creatorId: Long? = null,

    /**
     * 学校ID（租户标识）
     */
    @Column(name = "school_id")
    val schoolId: Long? = null,

    /**
     * 状态 (ACTIVE, INACTIVE, DELETED)
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20)
    val status: ExerciseBookStatus = ExerciseBookStatus.ACTIVE,

    /**
     * 练习册图片列表（JSON格式存储）
     */
    @Column(name = "images", columnDefinition = "JSON")
    val images: String? = null,

    /**
     * 创建时间
     */
    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    val createdAt: LocalDateTime? = null,

    /**
     * 更新时间
     */
    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    val updatedAt: LocalDateTime? = null
)

/**
 * 练习册状态枚举
 */
enum class ExerciseBookStatus {
    /**
     * 活跃状态
     */
    ACTIVE,

    /**
     * 非活跃状态
     */
    INACTIVE,

    /**
     * 已删除状态
     */
    DELETED
}

/**
 * 练习册图片信息
 */
data class ExerciseBookImage(
    /**
     * 图片URL
     */
    @JsonProperty("url")
    val url: String,

    /**
     * 图片描述
     */
    @JsonProperty("description")
    val description: String? = null,

    /**
     * 图片类型（封面图、内容图等）
     */
    @JsonProperty("type")
    val type: String? = "content",

    /**
     * 排序顺序
     */
    @JsonProperty("order")
    val order: Int? = 0
) {
    // 无参构造函数，用于Jackson反序列化
    constructor() : this("", null, "content", 0)
}

/**
 * ExerciseBookEntity 扩展属性和方法
 */

// 单例ObjectMapper，避免重复创建
private val objectMapper = com.fasterxml.jackson.databind.ObjectMapper()

/**
 * 获取图片列表
 */
val ExerciseBookEntity.imageList: List<ExerciseBookImage>
    get() = try {
        if (images.isNullOrBlank()) {
            emptyList()
        } else {
            objectMapper.readValue(
                images,
                object : com.fasterxml.jackson.core.type.TypeReference<List<ExerciseBookImage>>() {}
            )
        }
    } catch (e: Exception) {
        emptyList()
    }

/**
 * 获取封面图片URL
 */
val ExerciseBookEntity.coverImageUrl: String?
    get() = imageList.firstOrNull { it.type == "cover" }?.url
        ?: imageList.firstOrNull()?.url

/**
 * 获取所有图片URL列表
 */
val ExerciseBookEntity.imageUrls: List<String>
    get() = imageList.map { it.url }

/**
 * 获取班级ID列表
 */
val ExerciseBookEntity.classIdList: List<Long>
    get() = try {
        if (classIds.isNullOrBlank()) {
            // 如果classIds为空，但classId不为空，返回单个classId的列表
            classId?.let { listOf(it) } ?: emptyList()
        } else {
            objectMapper.readValue(
                classIds,
                object : com.fasterxml.jackson.core.type.TypeReference<List<Long>>() {}
            )
        }
    } catch (e: Exception) {
        // 解析失败时，尝试返回单个classId
        classId?.let { listOf(it) } ?: emptyList()
    }

/**
 * 获取班级名称列表
 */
val ExerciseBookEntity.classNameList: List<String>
    get() = try {
        if (classNames.isNullOrBlank()) {
            emptyList()
        } else {
            objectMapper.readValue(
                classNames,
                object : com.fasterxml.jackson.core.type.TypeReference<List<String>>() {}
            )
        }
    } catch (e: Exception) {
        emptyList()
    }

/**
 * 创建带图片的练习册实体
 */
fun ExerciseBookEntity.withImages(imageList: List<ExerciseBookImage>): ExerciseBookEntity {
    val imagesJson = if (imageList.isEmpty()) {
        null
    } else {
        try {
            objectMapper.writeValueAsString(imageList)
        } catch (e: Exception) {
            null
        }
    }

    return this.copy(images = imagesJson)
}

/**
 * 从URL列表创建图片信息
 */
fun createImagesFromUrls(urls: List<String>): List<ExerciseBookImage> {
    return urls.mapIndexed { index, url ->
        ExerciseBookImage(
            url = url,
            type = if (index == 0) "cover" else "content",
            order = index
        )
    }
}

/**
 * 创建带有班级ID列表的练习册实体
 */
fun ExerciseBookEntity.withClassIds(classIdList: List<Long>): ExerciseBookEntity {
    val classIdsJson = if (classIdList.isEmpty()) {
        null
    } else {
        try {
            objectMapper.writeValueAsString(classIdList)
        } catch (e: Exception) {
            null
        }
    }

    return this.copy(classIds = classIdsJson)
}

/**
 * 创建带有班级ID和名称列表的练习册实体
 */
fun ExerciseBookEntity.withClassIdsAndNames(classIdList: List<Long>, classNameList: List<String>): ExerciseBookEntity {
    val classIdsJson = if (classIdList.isEmpty()) {
        null
    } else {
        try {
            objectMapper.writeValueAsString(classIdList)
        } catch (e: Exception) {
            null
        }
    }

    val classNamesJson = if (classNameList.isEmpty()) {
        null
    } else {
        try {
            objectMapper.writeValueAsString(classNameList)
        } catch (e: Exception) {
            null
        }
    }

    return this.copy(classIds = classIdsJson, classNames = classNamesJson)
}